wordpress-mobile/WordPress-Android
# Create a Sequencing snackbar feature

Fixes #10771 and supersedes #10820 

The code in this PR was mostly done in Pair Programming/Brainstorming with @malinajirka , so credits to him 🙇‍♂️. 

The basic idea is to introduce a snackbar sequencer to avoid that multiple snackbars, triggered nearly together, overlap not allowing to the user to check/understand the ongoing messages. We introduced a SnackbarSequencer class and a SnackbarSequencerInfo class. The client code can enqueue SnackbarSequencerInfo element into the SnackbarSequencer instead of directly use the WPSnackbar class. In this way the sequencer consumes the queue of SnackbarSequencerInfo visualuazing in sequence the snackbars.

To avoid memory leaks due to the fact that we keep into the sequencer queue  references to view/context and listeners that can eventually be no more active at the time we try to show the snackbar, we keep those references as Weak/Soft references.

For now we have used the sequencer only in upload posts related snackbars to limit the impact as much as possible to the #10771 issue. It will be eventually extended in a future time to all snackbars. NOTE: only snackbars that are not `Snackbar.LENGTH_INDEFINITE` are allowed, otherwise an `IllegalArgumentException` is triggered.

For the unit testing we introduced `kotlinx-coroutines-test` version 1.2.1 in the testImplementation tree. In this way we can use the `advanceTimeBy` function to test the coroutine/delay.

Pinging also @planarvoid to have a third couple of eyes on it.

## To test:
- Use the instructions in the #10771 (or use Charles Proxy and its Block List feature) to trigger an upload fail and check that now the multiple snackbars are shown in sequence and you can open the snackbar actions in between
- (this is tricky to execute and describe tbh) Use the AS to start debugging WP app and attach a memory profiler session:
    - place a breakpoint in `SnackbarSequencer.start` (in `val item = snackBarQueue.peek()`)
    - Open PostsList > Drafts and trash a post
    - the above breakpoint is hit
    - in the app exit from PostsListActivity using the back button and going to the main activity
    - in the profiler check that the PostsListActivity is still in memory
    - resume debugger and check in the logcat that you get the message `SnackbarSequencer > start context was null` or `SnackbarSequencer > start context was not alive` (this means the snackbar got the Activity was no more valid)
    - in the memory profiler force some cycles of garbage collection (using the bin icon); and check the PostsListActivity gets garbaged (can need 10/12 cycles or maybe more depending on memory pressure)
- Unit testing must pass


PR submission checklist:
